File: src/sdk/typescript/src/utils/retry.ts
/**
 * Retry logic with exponential backoff
 */

export interface RetryConfig {
  maxRetries: number;
  initialDelay: number;
  maxDelay: number;
  backoffMultiplier: number;
}

export interface RetryOptions {
  maxRetries?: number;
  initialDelay?: number;
  maxDelay?: number;
  backoffMultiplier?: number;
  onRetry?: (attempt: number, delay: number, error: Error) => void;
}

/**
 * Calculate delay for retry attempt
 */
export function calculateDelay(
  attempt: number,
  initialDelay: number,
  maxDelay: number,
  backoffMultiplier: number
): number {
  const delay = initialDelay * Math.pow(backoffMultiplier, attempt - 1);
  return Math.min(delay, maxDelay);
}

/**
 * Sleep for specified milliseconds
 */
export function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Retry a function with exponential backoff
 */
export async function retry<T>(
  fn: () => Promise<T>,
  config: RetryConfig,
  options?: RetryOptions
): Promise<T> {
  const maxRetries = options?.maxRetries ?? config.maxRetries;
  const initialDelay = options?.initialDelay ?? config.initialDelay;
  const maxDelay = options?.maxDelay ?? config.maxDelay;
  const backoffMultiplier = options?.backoffMultiplier ?? config.backoffMultiplier;

  let lastError: Error | null = null;

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error instanceof Error ? error : new Error(String(error));

      if (attempt === maxRetries) {
        throw lastError;
      }

      const delay = calculateDelay(
        attempt + 1,
        initialDelay,
        maxDelay,
        backoffMultiplier
      );

      options?.onRetry?.(attempt + 1, delay, lastError);

      await sleep(delay);
    }
  }

  throw lastError;
}

/**
 * Check if error is retryable
 */
export function isRetryableError(error: unknown): boolean {
  if (!(error instanceof Error)) return false;

  const message = error.message.toLowerCase();

  // Network errors are retryable
  if (
    message.includes('econnrefused') ||
    message.includes('econnreset') ||
    message.includes('etimedout') ||
    message.includes('enotfound') ||
    message.includes('network error')
  ) {
    return true;
  }

  // 5xx errors are retryable
  if (message.includes('5') || message.includes('server error')) {
    return true;
  }

  // 429 (rate limit) is retryable
  if (message.includes('429') || message.includes('rate limit')) {
    return true;
  }

  // 503 (service unavailable) is retryable
  if (message.includes('503') || message.includes('service unavailable')) {
    return true;
  }

  return false;
}

/**
 * Create a retry configuration preset
 */
export const RetryPresets = {
  aggressive: {
    maxRetries: 5,
    initialDelay: 50,
    maxDelay: 5000,
    backoffMultiplier: 2
  } as RetryConfig,

  moderate: {
    maxRetries: 3,
    initialDelay: 100,
    maxDelay: 10000,
    backoffMultiplier: 2
  } as RetryConfig,

  conservative: {
    maxRetries: 1,
    initialDelay: 500,
    maxDelay: 5000,
    backoffMultiplier: 2
  } as RetryConfig,

  none: {
    maxRetries: 0,
    initialDelay: 0,
    maxDelay: 0,
    backoffMultiplier: 1
  } as RetryConfig
};

File: src/sdk/python/ai_agent/retry.py
"""
Retry logic with exponential backoff
"""

import asyncio
import logging
from typing import Awaitable, Callable, Optional, TypeVar, Any

T = TypeVar("T")


class RetryConfig:
    """Retry configuration"""

    def __init__(
        self,
        max_retries: int = 3,
        initial_delay: int = 100,
        max_delay: int = 10000,
        backoff_multiplier: float = 2.0,
    ):
        self.max_retries = max_retries
        self.initial_delay = initial_delay
        self.max_delay = max_delay
        self.backoff_multiplier = backoff_multiplier


def calculate_delay(
    attempt: int,
    initial_delay: int,
    max_delay: int,
    backoff_multiplier: float,
) -> int:
    """Calculate delay for retry attempt"""
    delay = int(initial_delay * (backoff_multiplier ** (attempt - 1)))
    return min(delay, max_delay)


async def retry(
    fn: Callable[..., Awaitable[T]],
    config: RetryConfig,
    on_retry: Optional[Callable[[int, int, Exception], None]] = None,
    *args: Any,
    **kwargs: Any,
) -> T:
    """Retry a function with exponential backoff"""
    logger = logging.getLogger(__name__)
    last_error: Optional[Exception] = None

    for attempt in range(config.max_retries + 1):
        try:
            return await fn(*args, **kwargs)
        except Exception as error:
            last_error = error

            if attempt == config.max_retries:
                raise last_error

            delay = calculate_delay(
                attempt + 1,
                config.initial_delay,
                config.max_delay,
                config.backoff_multiplier,
            )

            if on_retry:
                on_retry(attempt + 1, delay, error)

            logger.debug(
                f"Retry attempt {attempt + 1} in {delay}ms due to: {error}"
            )
            await asyncio.sleep(delay / 1000.0)

    raise last_error


def is_retryable_error(error: Exception) -> bool:
    """Check if error is retryable"""
    message = str(error).lower()

    # Network errors are retryable
    if any(
        x in message
        for x in ["connection refused", "connection reset", "timeout", "not found"]
    ):
        return True

    # 5xx errors are retryable
    if "5" in message or "server error" in message:
        return True

    # 429 (rate limit) is retryable
    if "429" in message or "rate limit" in message:
        return True

    # 503 (service unavailable) is retryable
    if "503" in message or "service unavailable" in message:
        return True

    return False


class RetryPresets:
    """Retry configuration presets"""

    AGGRESSIVE = RetryConfig(
        max_retries=5, initial_delay=50, max_delay=5000, backoff_multiplier=2.0
    )

    MODERATE = RetryConfig(
        max_retries=3, initial_delay=100, max_delay=10000, backoff_multiplier=2.0
    )

    CONSERVATIVE = RetryConfig(
        max_retries=1, initial_delay=500, max_delay=5000, backoff_multiplier=2.0
    )

    NONE = RetryConfig(
        max_retries=0, initial_delay=0, max_delay=0, backoff_multiplier=1.0
    )
